### ISSUE TITLE
FlashAttention3 propagates NaN from KVCache paddings to output

### ISSUE THREAD

**Issue by SpinDOS** [Dao-AILab/flash-attention#1745]
Hello! Thank for your great work.
I found an issue with FlashAttention3: it incorrectly processes unintialized data with possible NaN/inf inside paddings of KV cache at forward pass. As a result, presence of NaN/inf at paddings may result to incorrect NaN in the outputs. 

More details:
Python API has `flash_attn_with_kvcache` method, which accepts `k_cache`, `v_cache` and `cache_seqlens`. With FlashAttention2 it was a valid case to pass `v_cache` with uninitialized data at padding positions (`position >= cache_seqlens[batch_idx]`). FlashAttention2 has a logic to correctly ignore such an out-of-bounds data, so the output is not polluted with NaN. But with FlashAttention3 I observe different behaviour: NaN/inf inside KV cache padding does pollute attention output.

Please see the following repro:
<details>
  <summary>python repro</summary>

  ```
  import flash_attn as flash_2
  import flash_attn_interface as flash_3
  import torch
  
  batch_size = 2
  max_len = 3
  lens = torch.tensor([1, 3], dtype=torch.int32).cuda()
  q = torch.full((batch_size, max_len, 1, 64), 0.5, dtype=torch.float16).cuda()
  k = torch.full((batch_size, max_len, 1, 64), torch.nan, dtype=torch.float16).cuda()
  v = torch.full((batch_size, max_len, 1, 64), torch.nan, dtype=torch.float16).cuda()
  # page_table = torch.tensor([[0], [1]], dtype=torch.int32).cuda()
  
  for i, len in enumerate(lens):
      k[i, :len, :, :] = 0.5
      v[i, :len, :, :] = 0.5
  
  out = flash_2.flash_attn_with_kvcache(q, k, v, cache_seqlens=lens)
  print('flash_2', out.max().item())
  
  out = flash_3.flash_attn_with_kvcache(q, k, v, cache_seqlens=lens)
  print('flash_3', out.max().item())
  ```

  This code results in the following output:
  ```
  flash_2 0.5
  flash_3 nan
  ```

</details>

Interestingly, the problem disappears with paged_kv. If I add param `page_table=page_table` to the repro above, FlashAttention3 output does not contain NaN.

Is it a bug? Or KV Cache with NaN is considered as invalid input, so I observe legal UB? Should I wait for a fix in FlashAttention3, or fill padding on the caller side?

**Comment by tridao** [comment]
Thanks for the bug report. 
Indeed it's possible to get NaN if the uninitialized data has NaN/Inf in V (not in K). This is due to the fact that we use TMA.
I recommend zero'ing out the KV cache first.
We've been trying to think of a solution that can mask this out (e.g. don't use TMA, or use TMA and then explicitly mask out NaN values in V), but so far it looks like it might hurt performance.
We'll document this caveat, with the possibility that at some point we will be able to fix it.

**Comment by SpinDOS** [comment]
Okey, so I will zero out KV cache, or switch to paged version of KV cache.
Thanks.

**Comment by SpinDOS** [comment]
JFYI: switch to paged version of KV cache slows down attention by 10-20% in my benchmarks. So I prefer to zero out KV cache.
